server/chat-plugins/announcements.ts

```diff
@@ -42,7 +53,48 @@ export class Announcement {
 	}
 
 	end() {
+		this.endTimer();
 		this.room.send(`|uhtmlchange|announcement${this.announcementNumber}|<div class="infobox">(${this.room.tr`The announcement has ended.`})</div>`);
+		delete this.room.settings.minorActivity;
+		this.room.minorActivity = null;
+		this.room.saveSettings();
+	}
+	save() {
+		const entry: AnnouncementData = {
+			source: this.source,
+			announcementNumber: this.announcementNumber,
+			timeoutMins: this.timeoutMins,
+			timerEnd: this.timerEnd,
+			activityId: 'announcement',
+		};
+		this.room.settings.minorActivity = entry;
+		this.room.saveSettings();
+		return entry;
+	}
+	runTimeout(timeout: number) {
+		this.timeoutMins = timeout;
+		this.timerEnd = Date.now() + timeout * 60000;
+		this.timeout = setTimeout(() => {
+			const room = this.room;
+			if (!room) return; // do nothing if the room doesn't exist anymore
+			if (room.minorActivity) room.minorActivity.end();
+		}, timeout * MINUTE);
+		this.save();
+		return this.timeout;
+	}
+	endTimer() {
+		if (!this.timeout) return;
+		clearTimeout(this.timeout);
+		this.timeoutMins = 0;
+		delete this.timerEnd;
+		return this;
+	}
+}
+
+for (const room of Rooms.rooms.values()) { // hotpatching!
+	if (room.settings.minorActivity?.activityId === 'announcement') {
+		if (room.minorActivity?.timeout) clearTimeout(room.minorActivity.timeout);
```

Comment:
```suggestion
		room.minorActivity?.destroy();
```

Should probably be an interface looking more like this tbh.

Reply:
I don't think this needs to be a `destroy` function when you can just clear the timer, since that's the main issue here. 
`end` handles the other stuff that destroy might.

Reply:
Using `endTimer` here would be better, though.

Reply:
A `destroy` function is better futureproofing. You don't know exactly what a `minorActivity` needs to do to clean up, you're just generically telling it to clean up so it can be replaced.